Allow namenode as parameter instead of conf file.

asqasq opened the pull request:
In the serverless environment, the Crail client within the serverless function cannot access the
Crail configuration file, as there is no file system mounted on the function
execution environment.

This minor change allows the Crail client running as serverless function
to sepcify  the namenode address in the constructor instead of in the configuration file.

Signed-off-by: Adrian Schuepbach <asq@apache.org>

patrickstuedi commented on client/src/main/java/org/apache/crail/conf/CrailConfiguration.java:
```diff
@@ -39,6 +39,17 @@ public CrailConfiguration() throws IOException {
+  // read the namenode address from any configuration file.
+  public CrailConfiguration(String nameNodeAddress) throws IOException {
+    conf = new ConcurrentHashMap<>();
+    conf.put("crail.namenode.address", nameNodeAddress);
```
I don't mind adding this line. But this really only addresses the namenode parameter (obviously a key parameter). Would be nice to have a more flexible solution where either we can load the config also from different sources, or allow the user to create a plain config and then set certain properties by hand in the case where CRAIL_HOME is not set and properties cannot be loaded.

asqasq replied:
Thanks for the feedback.

I agree, it is a simplistic change and a more generic solution would definitely be better.
One thing I could do is starting with this simple change for now and at the same time create and assign a ticket to myself and come up with a more generic solution.

PepperJo replied:
Why not instead make a constructor which creates an empty configuration? and the client can then set all the configuration parameters to its liking?

asqasq replied:
I would be fine with this idea as well.

The easiest way would be to keep the already existing constructor CrailConfiguration(), but move
the call of loadProperties into a public function, which can optionally be called by the user,
if hte wants to read the configuration file.

If the user choses not to load the configuration file, he could just use the existing set() method.

If this is fine with everyone, I will change it accordingly.

PepperJo replied:
Sounds good. Alternatively a static function: createFromConfigurationFile() or similar, but I don't have a strong opinion about either.

patrickstuedi replied:
Anything is fine as long as we are not breaking existing applications and benchmarks that assume the properties to be loaded when calling the empty constructor.

asqasq replied:
Good point, actually I was thinking about that as well... If we change it this way, the default empty  constructor would not load the properties, which will break existing code. This is why I originally introduced a different (simplistic) constructor. With this one we would not have the problem of breaking or needing to change existing code.

PepperJo replied:
This works both ways though. We can leave the constructor as is but introduce a new static public method `createEmtpy` or similar.
Anyway I personally am not to concerned about breaking things we have done this numerous times in the past. New version, new API is fine for me.

asqasq commented:
One way to have more flexibility and be semantically consistent,
would be to introduce three static functions to create an
empty configuration, create a configuration from the default
configuration file and create a configuration from the file given
as parameter and set the current public constructor to private.

The code would be something like this:

```
private CrailConfiguration() {
  ...
}

public static CrailConfiguration createEmptyConfiguration() {
  ...
}
public static CrailConfiguration createConfigurationFromFile() {
  ...
}
public static CrailConfiguration createConfigurationFromFile(String path) {
  ...
}
```

The function `createEmptyConfiguration` instantiates an empty configuration.
The user can set properties by calling `public void set(String key, String value)`
and `public void setInt(String key, int level)`, two already existing functions.

`createConfigurationFromFile` instantiates `CrailConfiguration` and
loads the default configuration file. This would be the exact same
behaviour as the current default public constructior `CrailConfiguration()`.

`createConfigurationFromFile(String path)` would do the same as the above
function, but read the configuration from the explicitely given full
path to the configuration file, instead of loading the default
configuration file.

This implies a change to the applications. Applications cannot
call `new CrailConfiguration()` anymore (constructor is private),
but need to be changed to call `createConfigurationFromFile()`,
which has the exact same semantic as the currently public
`CrailConfiguration()` constructor.

How does that sounds?

patrickstuedi commented:
Looks good!

asqasq commented:
Hi all

I implemented the functions as proposed above and updated the PR. I tested the following:
- Regular Crail based on TCP works (namenode, datanode, Crail client)
- Containerized Crail on Kubernetes works (namenode, datanode, NVMf datanode, client)
- Serverless function on OpenWhisk can access containerized Crail service)
- fsck works
- iobench (tested with -t read and -t write) works
- binary SNAPSHOT (.tar.gz) produced by Maven works

Could you please have a look at the updated commit?
Thanks

PepperJo commented on client/src/main/java/org/apache/crail/conf/CrailConfiguration.java:
```diff
@@ -33,9 +33,32 @@
+		if (base == null || base.isEmpty()) {
+			throw new IllegalArgumentException("CRAIL_HOME environment variable is not set or empty");
+		}
+		cconf.loadConfigurationFromFile(base + "/conf/crail-site.conf");
```
I recommend using File.separator instead of '/', makes it more portable.

asqasq replied:
Good idea!

Done.
